### PR TITLE
Add cas command

### LINKED ISSUE #47: Support CAS commands

### PR DESCRIPTION
**Pull request by letmutx**
Fixes #47 

Haven't yet tested, I'll add tests soon.

### CODE REVIEW

**Comment by letmutx** [comment]
@aisk Please review when you get a chance!

**Comment by aisk** [comment]
Thanks @letmutx ! I'll review this as soon as posible!

**Comment by letmutx** [comment]
@aisk Just a heads up - please do not merge this yet. I added more tests and a particular test is failing. I am debugging. Will try and fix over the weekend.

**Comment by letmutx** [comment]
@aisk Fixed now. Should be good to review.

**Comment by letmutx** [comment]
The failing test is because of a debug_assert! I added in `parse_header_only_response`. In some cases, a body can follow after a `store` command. (for ex: when using cas, add, replace commands, if the key doesn't exist, an error string follows after the response header)

There are no tests for these cases. We should handle them in a different issue. I will remove the debug_assert and add an issue.

**Comment by aisk** [comment]
Hi sorry late for the review, I'm a little busy these days.

File: src/protocol/mod.rs [hunk 1]
```diff
@@ -9,7 +9,9 @@ pub(crate) use protocol::ascii::AsciiProtocol;
 pub(crate) use protocol::binary::BinaryProtocol;
 use std::collections::HashMap;
 use stream::Stream;
-use value::{FromMemcacheValue, ToMemcacheValue};
+use value::{FromMemcacheValueExt, ToMemcacheValue};
+
+pub type CasId = u64;
```
Comment (aisk) [inline src/protocol/mod.rs]: Seems `CasId` is just a `u64` alias, I think we should keep it simple and let users just using the native `u64` type, and to keep a unified API style while the `set`'s expiration is just an `u32`.

**Comment by letmutx** [comment]
The `Ext` trait is to keep backwards compatibility. Because the definition of `from_memcache_value` changes in the new trait, users who gave implemented the trait for their custom types would've had to implement the new trait.

But because there is a blanket impl of `FromMemcacheValueExt`for all types which implement `FromMemcacheValue`, they no longer need to do that.

File: src/protocol/mod.rs (hunk 1 above)
Comment (letmutx) [reply src/protocol/mod.rs]: I agree with you that keeping a consistent API is important but I think `CasId` is more descriptive than `u64`. Users can still use `u64` without any extra boilerplate.

**Comment by aisk** [comment]
Yes you are right, sorry I don't considered the backwards compatibility issue. We shoul keep it, thanks @letmutx

File: src/protocol/mod.rs (hunk 1 above)
Comment (aisk) [reply src/protocol/mod.rs]: I think we should keep it as a native `u64`, and maybe add a trait for this and expire param someday.

**Comment by aisk** [comment]
Hi @letmutx I approved this PR, except I want to remove the CasID type as mentioned above. I think maybe we should add a type or trait some day if we have a chance to refactor the API.

And can you rebase the master branch since there are some changes. Seems there was no conflict but I think a rebase with CI test is needed.

Really thanks @letmutx :+1:

**Comment by letmutx** [comment]
Thanks for the review.

**Comment by letmutx** [comment]
@aisk rebased and pushed the changes.

**Comment by aisk** [comment]
@letmutx merged, thanks!

**Pull request merged**